app/src/main/kotlin/edu/stanford/bdh/engagehf/questionnaire/QuestionnaireViewModel.kt

```diff
@@ -0,0 +1,109 @@
+package edu.stanford.bdh.engagehf.questionnaire
+
+import android.os.Bundle
+import androidx.core.os.bundleOf
+import androidx.lifecycle.SavedStateHandle
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import ca.uhn.fhir.parser.IParser
+import dagger.hilt.android.lifecycle.HiltViewModel
+import edu.stanford.spezi.core.logging.speziLogger
+import edu.stanford.spezi.core.navigation.NavigationEvent
+import edu.stanford.spezi.core.navigation.Navigator
+import edu.stanford.spezi.core.utils.MessageNotifier
+import edu.stanford.spezi.core.utils.extensions.decode
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import org.hl7.fhir.r4.model.QuestionnaireResponse
+import java.util.Date
+import javax.inject.Inject
+
+@HiltViewModel
+class QuestionnaireViewModel @Inject internal constructor(
+    private val questionnaireRepository: QuestionnaireRepository,
+    private val navigator: Navigator,
+    savedStateHandle: SavedStateHandle,
+    private val notifier: MessageNotifier,
+    private val jsonParser: IParser,
+) : ViewModel() {
+    private val logger by speziLogger()
+
+    private val _uiState = MutableStateFlow<State>(
+        State.Loading(
+            savedStateHandle.decode(QUESTIONNAIRE_SAVE_STATE_PARAM)
+        )
+    )
+    val uiState = _uiState.asStateFlow()
+
+    init {
+        loadQuestionnaire()
+    }
+
+    private fun loadQuestionnaire() {
+        viewModelScope.launch {
+            val questionnaireId = (_uiState.value as? State.Loading)?.questionnaireId
+            questionnaireId?.let {
+                val questionnaire =
+                    questionnaireRepository.observe(it).first()
```

Comment:
If we are just interested in the first emission, is it really required to have observe on repository? Isn't it sufficient to simply get once the questionnaire by id and return a kotlin.Result?

It would be great to also handle error case as well. currently the app would crash in case something goes wrong with internet connection, no questionnaire from firebase or if encoding fails

Reply:
Yea this makes a lot of sense. I changed it.